### PR TITLE
Refactor cupy.core and tests

### PR DESCRIPTION
**Pull request by okuta**

### CODE REVIEW

File: tests/cupy_tests/core_tests/test_core.py [hunk 1]
```diff
@@ -1,42 +1,9 @@
 import unittest
 
-import numpy
-
 import cupy
 from cupy.core import core
 
 
-class TestGetSize(unittest.TestCase):
```
Comment (delta2323) [inline tests/cupy_tests/core_tests/test_core.py]: It seems this unittest is removed. Could you restore it?

File: tests/cupy_tests/core_tests/test_core.py [hunk 2]
```diff
@@ -1,42 +1,9 @@
...
-
-    def test_int(self):
-        self.assertEqual(core.get_size(1), (1,))
-
-    def test_invalid(self):
-        with self.assertRaises(ValueError):
-            core.get_size(1.0)
-
-
-class TestInternalProd(unittest.TestCase):
```
Comment (delta2323) [inline tests/cupy_tests/core_tests/test_core.py]: ditto(Restore unittest).

File: tests/cupy_tests/core_tests/test_core.py [hunk 3]
```diff
@@ -52,52 +19,3 @@ def test_one(self):
     def test_normal(self):
         # TODO(nno): write test for normal case
         pass
-
-
-class TestGetContiguousStrides(unittest.TestCase):
```
Comment (delta2323) [inline tests/cupy_tests/core_tests/test_core.py]: ditto(Restore unittest).

File: tests/cupy_tests/core_tests/test_core.py [hunk 4]
```diff
@@ -52,52 +19,3 @@ def test_one(self):
...
-
-    def test_two(self):
-        self.assertEqual(core._get_contiguous_strides((1, 2), 3), [6, 3])
-
-    def test_three(self):
-        self.assertEqual(core._get_contiguous_strides((1, 2, 3), 4),
-                         [24, 12, 4])
-
-
-class TestGetCContiguity(unittest.TestCase):
```
Comment (delta2323) [inline tests/cupy_tests/core_tests/test_core.py]: ditto(Restore unittest).

File: tests/cupy_tests/core_tests/test_core.py [hunk 5]
```diff
@@ -52,52 +19,3 @@ def test_one(self):
...
-
-    def test_zero_in_shape(self):
-        self.assertTrue(core._get_c_contiguity((1, 0, 1), (1, 1, 1), 3))
-
-    def test_normal(self):
-        # TODO(unno): write test for normal case
-        pass
-
-
-class TestInferUnknownDimension(unittest.TestCase):
```
Comment (delta2323) [inline tests/cupy_tests/core_tests/test_core.py]: ditto(Restore unittest).

File: tests/cupy_tests/test_internal.py [hunk 6]
```diff
@@ -1,60 +0,0 @@
...
-    {'slice': (8, -3, -1),   'expect': (8, 7, -1)},
-
-    {'slice': (None, 8, -1),  'expect': (9, 8, -1)},
-    {'slice': (-3, 6, -1),    'expect': (7, 6, -1)},
-    {'slice': (11, 8, -1),    'expect': (10, 8, -1)},
-    {'slice': (11, 11, -1),   'expect': (10, 10, -1)},
-    {'slice': (-11, 8, -1),   'expect': (0, 0, -1)},
-    {'slice': (-11, -11, -1), 'expect': (0, 0, -1)},
-)
-class TestCompleteSlice(unittest.TestCase):
```
Comment (delta2323) [inline tests/cupy_tests/test_internal.py]: ditto(Restore unittest).

File: tests/cupy_tests/test_internal.py [hunk 7]
```diff
@@ -1,60 +0,0 @@
...
-
-    def test_complete_slice(self):
-        self.assertEqual(
-            core.complete_slice(slice(*self.slice), 10),
-            slice(*self.expect))
-
-
-class TestCompleteSliceError(unittest.TestCase):
```
Comment (delta2323) [inline tests/cupy_tests/test_internal.py]: ditto(Restore unittest).

File: cupy/core/internal.pyx [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+        return ()
+    if cpython.PySequence_Check(size):
+        return tuple(size)
+    if isinstance(size, int):
+        return size,
+    raise ValueError('size should be None, collections.Sequence, or int')
+
+
+@cython.profile(False)
+cpdef inline bint vector_equal(
```
Comment (delta2323) [inline cupy/core/internal.pyx]: Add unittest for `vector_equal`

File: cupy/core/internal.pyx [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+@cython.profile(False)
+cpdef inline Py_ssize_t prod(args, Py_ssize_t init=1) except *:
+    cdef Py_ssize_t arg
+    for arg in args:
+        init *= arg
+    return init
+
+
+@cython.profile(False)
+cpdef inline Py_ssize_t prod_ssize_t(
```
Comment (delta2323) [inline cupy/core/internal.pyx]: Add unittest for `prod_ssize_t`

File: cupy/core/internal.pyx [hunk 10]
```diff
@@ -0,0 +1,174 @@
+cimport cpython
+cimport cython
+
+@cython.profile(False)
+cpdef inline Py_ssize_t prod(args, Py_ssize_t init=1) except *:
```
Comment (delta2323) [inline cupy/core/internal.pyx]: Add unittest for `prod`

File: cupy/core/internal.pyx [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+    cdef Py_ssize_t n = x.size()
+    if n != y.size():
+        return False
+    for i in range(n):
+        if x[i] != y[i]:
+            return False
+    return True
+
+
+cpdef void get_reduced_dims(
```
Comment (delta2323) [inline cupy/core/internal.pyx]: Add unittest for `get_reduced_dims`

File: cupy/core/core.pyx [hunk 12]
```diff
@@ -149,7 +129,7 @@ cdef class ndarray:
 
         def __set__(self, newshape):
             cdef vector.vector[Py_ssize_t] shape, strides
-            shape = _infer_unknown_dimension(newshape, self.size)
+            shape = internal.infer_unknown_dimension(newshape, self.size)
             strides = _get_strides_for_nocopy_reshape(self, newshape)
```
Comment (delta2323) [inline cupy/core/core.pyx]: Why `_get_strides_for_nocopy_reshape` is not moved to `internal` module?

**Comment by delta2323** [comment]
@okuta  Thank you for applying my comments.
Now I come to think adding new test cases is beyond refactoring. So I merge this issue and create new issue for new test cases.

File: tests/cupy_tests/core_tests/test_core.py (hunk 1 above)
Comment (okuta) [inline tests/cupy_tests/core_tests/test_core.py]: Test is moved to `test_internal.py`.

File: tests/cupy_tests/core_tests/test_core.py (hunk 2 above)
Comment (okuta) [inline tests/cupy_tests/core_tests/test_core.py]: Test is moved to `test_internal.py`.

File: tests/cupy_tests/core_tests/test_core.py (hunk 3 above)
Comment (okuta) [inline tests/cupy_tests/core_tests/test_core.py]: Test is moved to `test_internal.py`.

File: tests/cupy_tests/core_tests/test_core.py (hunk 4 above)
Comment (okuta) [inline tests/cupy_tests/core_tests/test_core.py]: Test is moved to `test_internal.py`.

File: tests/cupy_tests/core_tests/test_core.py (hunk 5 above)
Comment (okuta) [inline tests/cupy_tests/core_tests/test_core.py]: Test is moved to `test_internal.py`.

File: tests/cupy_tests/test_internal.py (hunk 6 above)
Comment (okuta) [inline tests/cupy_tests/test_internal.py]: Test is moved to `test_internal.py`.

File: tests/cupy_tests/test_internal.py (hunk 7 above)
Comment (okuta) [inline tests/cupy_tests/test_internal.py]: Test is moved to `test_internal.py`.

File: cupy/core/core.pyx (hunk 12 above)
Comment (okuta) [inline cupy/core/core.pyx]: `_get_strides_for_nocopy_reshape` takes `cupy.ndarray` type argument.

**Comment by delta2323** [comment]
LGTM :+1:

File: cupy/core/core.pyx (hunk 12 above)
Comment (delta2323) [inline cupy/core/core.pyx]: OK, I understand it.

File: cupy/core/internal.pyx (hunk 9 above)
Comment (okuta) [inline cupy/core/internal.pyx]: I added unittest.

File: cupy/core/internal.pyx (hunk 8 above)
Comment (okuta) [inline cupy/core/internal.pyx]: I added unittest.

**Comment by okuta** [comment]
I added unittests.
Please check.

**Comment by delta2323** [comment]
I forgot to tell you to add unit tests to `get_reduced_dims`. Could you add them?

**Comment by okuta** [comment]
I can't add unit test for `get_reduced_dims`.
Because `get_reduced_dims` is c-language function.

**Comment by delta2323** [comment]
OK, I agree with you about `get_reduced_dims`.

**Comment by delta2323** [comment]
Could resolve the conflict?

**Comment by delta2323** [comment]
LGTM :+1:

**Pull request merged**